nemo_curator/image/classifiers/base.py

```diff
@@ -0,0 +1,112 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+from abc import ABC, abstractmethod
+from typing import Union
+
+import cupy as cp
+import torch
+
+from nemo_curator.datasets import ImageTextPairDataset
+from nemo_curator.utils.cudf_utils import create_list_series_from_1d_or_2d_ar
+from nemo_curator.utils.distributed_utils import load_object_on_worker
+
+
+class ImageClassifier(ABC):
+    """
+    An abstract base class that represents a classifier on top
+    of embeddings generated by a CLIP vision encoder
+    """
+
+    def __init__(
+        self,
+        model_name: str,
+        embedding_column: str,
+        pred_column: str,
+        pred_type: Union[str, type],
+        batch_size: int,
+        embedding_size: int,
+    ) -> None:
+        self.model_name = model_name
+        self.embedding_column = embedding_column
+        self.pred_column = pred_column
+        self.pred_type = pred_type
+        self.batch_size = batch_size
+        self.embedding_size = embedding_size
+
+    def __call__(self, dataset: ImageTextPairDataset) -> ImageTextPairDataset:
+        meta = dataset.metadata.dtypes.to_dict()
+        meta[self.pred_column] = self.pred_type
+        embedding_df = dataset.metadata.map_partitions(self._run_inference, meta=meta)
+
+        return ImageTextPairDataset(
+            dataset.path,
+            metadata=embedding_df,
+            tar_files=dataset.tar_files,
+            id_col=dataset.id_col,
+        )
+
+    def _run_inference(self, partition, partition_info=None):
+        device_id = int(os.environ["CUDA_VISIBLE_DEVICES"].split(",")[0])
+        device = f"cuda:{device_id}"
+
+        model = load_object_on_worker(
+            self.model_name,
+            self.load_model,
+            {"device": device},
+        )
+
+        embeddings = torch.as_tensor(
```

Comment:
@VibhuJawa I'll address the rest of your comments later, but I noticed that there is actually a bug on this line. Printing `partition[self.embedding_column][:5]` right before this (on line 69) gives me this:

```
0    [-0.01650974154472351, 0.03625061735510826, -0...
1    [-0.01650974154472351, 0.03625061735510826, -0...
2    [-0.01650974154472351, 0.03625061735510826, -0...
3    [-0.01650974154472351, 0.03625061735510826, -0...
4    [0.021344991400837898, 0.03247494995594025, 0....
Name: image_embedding, dtype: list
```

While printing out `embeddings[:5]` after this gives me this:
```
tensor([[0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.],
        [0., 0., 0.,  ..., 0., 0., 0.]], device='cuda:1')
```

Any idea what could be going on here?

Reply:
I can send you the exact data if you want, but the minimum reproducer seems to be this:

```python
import os
from nemo_curator import get_client
from nemo_curator.datasets import ImageTextPairDataset
import dask_cudf
import torch

def _map_part(df):
    device_id = int(os.environ["CUDA_VISIBLE_DEVICES"].split(",")[0])
    device = f"cuda:{device_id}"
    print(device)
    print(df["image_embedding"][:5])

    embeddings = torch.as_tensor(
        df["image_embedding"].list.leaves.values.reshape(
            len(df), -1
        ),
        device=device,
    )
    print(embeddings[:5])
    return df

def main():
    client = get_client(cluster_type="gpu")
    input_dir = "/home/rywolf/Documents/experiments/image/cudf-test"

    id_col = "key"

    dataset = ImageTextPairDataset.from_webdataset(input_dir, id_col)
    dataset.metadata = dataset.metadata[dataset.metadata["error_message"].isna()]
    meta = dataset.metadata.dtypes.to_dict()
    files = dataset.metadata.map_partitions(_map_part, meta=meta)
    files.head()

if __name__ == "__main__":
    main()
```

This bug is super sporadic though. It only happens some of the time, and it appears that it only happens when the device is `cuda:1`, for `cuda:0` the conversion to a tensor is totally fine.

Reply:
Setting the `device="cuda"` seems to fix the error, but now it shows up that the tensor is on `cuda:0` when the worker should be assigned to `cuda:1`.

```python
import os
from nemo_curator import get_client
from nemo_curator.datasets import ImageTextPairDataset
import dask_cudf
import torch

def _map_part(df):
    device_id = int(os.environ["CUDA_VISIBLE_DEVICES"].split(",")[0])
    device = f"cuda:{device_id}"
    print(device)
    print(df["image_embedding"][:5])

    embeddings = torch.as_tensor(
        df["image_embedding"].list.leaves.values.reshape(
            len(df), -1
        ),
        # device=device,
        device="cuda",
    )
    print(embeddings[:5])
    return df

def main():
    client = get_client(cluster_type="gpu")
    input_dir = "/home/rywolf/Documents/experiments/image/cudf-test"

    id_col = "key"

    dataset = ImageTextPairDataset.from_webdataset(input_dir, id_col)
    dataset.metadata = dataset.metadata[dataset.metadata["error_message"].isna()]
    meta = dataset.metadata.dtypes.to_dict()
    files = dataset.metadata.map_partitions(_map_part, meta=meta)
    files.head()

if __name__ == "__main__":
    main()
```

Output:
```
cuda:1
0    [-0.01650974154472351, 0.03625061735510826, -0...
1    [-0.01650974154472351, 0.03625061735510826, -0...
2    [-0.01650974154472351, 0.03625061735510826, -0...
3    [-0.01650974154472351, 0.03625061735510826, -0...
4    [0.021344991400837898, 0.03247494995594025, 0....
Name: image_embedding, dtype: list
tensor([[-0.0165,  0.0363, -0.0122,  ..., -0.0325,  0.0112,  0.0216],
        [-0.0165,  0.0363, -0.0122,  ..., -0.0325,  0.0112,  0.0216],
        [-0.0165,  0.0363, -0.0122,  ..., -0.0325,  0.0112,  0.0216],
        [-0.0165,  0.0363, -0.0122,  ..., -0.0325,  0.0112,  0.0216],
        [ 0.0213,  0.0325,  0.0027,  ..., -0.0159, -0.0037,  0.0003]],
       device='cuda:0')
```

Reply:
Update: we solved this by no longer manually specifying the device id.